lib/three_scale/middleware/presigned_downloads.rb

```diff
@@ -0,0 +1,88 @@
+# frozen_string_literal: true
+
+module ThreeScale
+  module Middleware
+    class PresignedDownloads
+      RE_SEPARATOR = ::File::SEPARATOR == "/" ? "/" : "[#{Regexp.escape(::File::SEPARATOR)}/]"
+
+      delegate *%i[requires_signing? good_signature? signature_at to_time], to: :class, private: true
+
+      class << self
+        def verifier
+          Rails.application.message_verifier("file downloads")
+        end
+
+        def requires_signing?(path)
+          # for the used separator see https://github.com/rack/rack/pull/2257
+          # question mark on first separator is paranoia of getting a relative path somehow
+          invoices_re = %r{#{RE_SEPARATOR}?system#{RE_SEPARATOR}.*#{RE_SEPARATOR}invoices#{RE_SEPARATOR}}
+          path.start_with? invoices_re
+        end
+
+        def sign_path_if_needed(path, expires_in = 3600)
+          requires_signing?(path) ? sign(path, expires_in) : path
+        end
+
+        def sign(path, expires_in = 3600)
+          sign_at(path, Time.now.utc.round + expires_in)
+        end
+
+        def sign_at(path, expires_at)
+          expires_timestamp = to_timestamp(expires_at)
+          signature = signature_at(path, expires_at)
+          params = { "3scale-Expires" => expires_timestamp, "3scale-Signature" => signature }.to_param
+
+          "#{path}?#{params}"
+        end
+
+        def signature_at(path, expires_at)
+          signing_options = { purpose: :download, expires_at: }
+          verifier.generate(path.b, **signing_options).split("--").last
```

Comment:
why do we discard everything before the last `--`, what's in there?

Reply:
See https://api.rubyonrails.org/classes/ActiveSupport/MessageVerifier.html


Before `--` is the actual message that we sign. But it contains only redundant information that we already know form request path and expiration parameter. To make URLs shorter and more user friendly, I kept only the signature part.

Technically we can omit the expiration parameter and keep the whole signed message (which is just base64 encoded json), but I think it is more user friendly for the user to see the URL in this way and know when it is expiring in a human readable form.